src/zenml/orchestrators/utils.py

```diff
@@ -102,10 +103,279 @@ def is_setting_enabled(
+    """Get the return targets for this request.
+
+    Returns:
+        Mapping of step_name -> expected_output_name (or None for first)
+    """
+    return _return_targets.get({})
+
+
+def extract_return_contract(
+    pipeline_source: Optional[str],
+) -> Optional[Dict[str, str]]:
+    """Extract which step outputs to return from pipeline function using AST parsing.
+
+    Args:
+        pipeline_source: Source path of pipeline function
+
+    Returns:
+        Dict mapping output names to step names, or None if extraction fails
+    """
+    if not pipeline_source:
+        return None
+
+    try:
+        import ast
+        import inspect
+
+        from zenml.utils import source_utils
+
+        # Load pipeline function
+        pipeline_func = source_utils.load(pipeline_source)
+        source_code = inspect.getsource(pipeline_func)
+
+        # Parse source code into AST
+        tree = ast.parse(source_code)
+
+        # Find the function definition
+        func_def = None
+        for node in ast.walk(tree):
+            if isinstance(node, ast.FunctionDef):
+                func_def = node
+                break
+
+        if not func_def:
+            return None
+
+        # Find return statements and extract the last one
+        return_nodes = []
+        for node in ast.walk(func_def):
+            if isinstance(node, ast.Return) and node.value is not None:
+                return_nodes.append(node)
+
+        if not return_nodes:
+            return None
+
+        # Analyze the last return statement
+        last_return = return_nodes[-1]
+        return_vars = _extract_return_variables(last_return.value)
+
+        if return_vars:
+            # Map output names to step names
+            return {f"output_{i}": var for i, var in enumerate(return_vars)}
+
+    except Exception:
+        # If extraction fails, fall back gracefully
+        pass
+
+    return None
+
+
+def _extract_return_variables(node: Any) -> List[str]:
+    """Extract variable names from a return statement AST node.
+
+    Args:
+        node: AST node representing the return expression
+
+    Returns:
+        List of variable names being returned
+    """
+    import ast
+
+    if isinstance(node, ast.Name):
+        # Simple variable: return a
+        return [node.id]
+    elif isinstance(node, ast.Tuple):
+        # Tuple: return (a, b) or return a, b
+        vars_list = []
+        for elt in node.elts:
+            if isinstance(elt, ast.Name):
+                vars_list.append(elt.id)
+            else:
+                # For complex expressions, use a fallback name
+                vars_list.append(f"expr_{len(vars_list)}")
+        return vars_list
+    elif isinstance(node, ast.List):
+        # List: return [a, b]
+        vars_list = []
+        for elt in node.elts:
+            if isinstance(elt, ast.Name):
+                vars_list.append(elt.id)
+            else:
+                vars_list.append(f"expr_{len(vars_list)}")
```

Comment:
how does this `expr_` something name even help you detect that from the step outputs?

Reply:
outdated